Find a functionally equivalent code:com.android.keyguard.KeyguardHostView.ensureTransportPresentOrRemoved:COMMENT
Method Modifier: private     
Comment:/**
 * Examines the current state and adds the transport to the widget pager when the state changes.
 *
 * Showing the initial transport and keeping it around is a bit tricky because the signals
 * coming from music players aren't always clear. Here's how the states are handled:
 *
 * {@link TRANSPORT_GONE} means we have no reason to show the transport - remove it if present.
 *
 * {@link TRANSPORT_INVISIBLE} means we have potential to show the transport because a music
 * player is registered but not currently playing music (or we don't know the state yet). The
 * code adds it conditionally on play state.
 *
 * {@link #TRANSPORT_VISIBLE} means a music player is active and transport should be showing.
 *
 * Once the transport is showing, we always show it until keyguard is dismissed. This state is
 * maintained by onSave/RestoreInstanceState(). This state is cleared in
 * {@link KeyguardViewManager#hide} when keyguard is dismissed, which causes the transport to be
 * gone when keyguard is restarted until we get an update with the current state.
 *
 * @param state
 */

@@ -3,10 +3,8 @@
     final boolean visible = state == TRANSPORT_VISIBLE;
     final boolean shouldBeVisible = state == TRANSPORT_INVISIBLE && isMusicPlaying(state);
     if (!showing && (visible || shouldBeVisible)) {
-        if (DEBUGXPORT)
-            Log.v(TAG, "add transport");
         // insert to left of camera if it exists, otherwise after right-most widget
         int lastWidget = mAppWidgetContainer.getChildCount() - 1;
         // handle no widget case
@@ -12,18 +10,19 @@
         if (lastWidget >= 0) {
             position = mAppWidgetContainer.isCameraPage(lastWidget) ? lastWidget : lastWidget + 1;
         }
+        if (DEBUGXPORT)
+            Log.v(TAG, "add transport at " + position);
         mAppWidgetContainer.addWidget(getOrCreateTransportControl(), position);
+        return true;
     } else if (showing && state == TRANSPORT_GONE) {
         if (DEBUGXPORT)
             Log.v(TAG, "remove transport");
@@ -20,23 +21,25 @@
         mTransportControl = null;
         KeyguardUpdateMonitor.getInstance(getContext()).dispatchSetBackground(null);
     }
+    return false;
 }

